enhancements/installer/gcp-ipi-shared-vpc.md

```diff
@@ -0,0 +1,262 @@
+### Goals
+
+- Enable users to deploy GCP clusters with Shared VPC networking using the IPI
+  workflow.
+- Enable users to provide service account(s) with sufficient permissions in the
+  Host Project to enable provisioning the majority of the cluster resources.
+- Enable users to alternatively provide pre-provisioned resources to enable
+  minimal permission requirements in the Host Project.
+
+### Non-Goals
+
+- Provisioning and/or configuring the VPC, Network, and Subnetworks are to be
+  done manually by the user prior to running the installer.
+- Provisioning and/or configuring the Service Accounts for the Installer and
+  Cluster are to be done manually by the user prior to running the installer.
+
+## Proposal
+
+The installer should be able to install a GCP IPI cluster using Shared VPC
+networking. This requires some resources to be created and configured in
+advance, such as the network and subnets. Other resources can be created in
+advance when the necessary host project permissions are not desired, such as
+the firewall rules and/or DNS configuration. The installer, and some operators,
+will be updated to enable this functionality. They will become capable of
+creating and managing resources that exist in a different (Host) project from
+the cluster (Service Project). They will also become capable of consuming
+pre-existing firewall and/or DNS configurations.
+
+### Workflow Description
+
+Prior to executing the installer, the Shared VPC, Network, and Subnets will be
+manually created in the Host Project and configured for Shared VPC (XPN)
+configuration with the Service Project. A service account will also be created
+in the Service Project and granted sufficient permissions in the Host Project.
+When minimal permissions are desired in the Host Project, additional resources
+will be manually created and configured in the Host Project to include:
+firewall rules, api compute address, ingress compute address, and the api,
+api-int, and \*.apps DNS records.
+
+After all of these resources are created, they will be used in a manually
+created install-config.yaml to configure the cluster accordingly. At a minimum,
+the credentialsMode will be set to Passthrough, and the network,
+computeNetwork, controlPlaneNetwork, and networkProjectID will be configured to
+reference the Shared VPC network. When minimial permissions are desired in the
+Host Project, the additional parameters will be configured to enable the
+cluster to consume them as well.
+
+The installation process can then proceed as any other IPI workflow by running
+the create manifests, create ignition, and/or create cluster sub-commands.
+
+### API Extensions
+
+#### api.dns
+
+The current DNSZone struct does not appear to be able to store which project
+the zone is in. The ID variable can be used to store fully qualified path
+names, which will include the project. However, this may require modification
+of all consumers of this variable to be able to parse this information.
```

Comment:
It looks like the [current behavior for the cluster ingress operator is](https://github.com/openshift/cluster-ingress-operator/blob/master/pkg/operator/controller/dns/controller.go#L563) to use the project from [the cluster infrastructure object](https://github.com/openshift/api/blob/master/config/v1/types_infrastructure.go#L452).

We probably need to add a field to that object.

Reply:
I agree. I intended to enable the ingress operator developers to determine the field specifics.

Reply:
Just in case it's useful, for MAPI we added a field to the network interface struct, if a user sets this to a different project ID than that of the Machine, we create the NIC in the other project, https://github.com/openshift/api/blob/6c4f186339a7a4f505e133e9b86acd0d5aa84686/machine/v1beta1/types_gcpprovider.go#L134

Reply:
I believe the instance code is already good go go in regard to networking. The majority of the effort here is about the firewalls, DNS, and load balancers by extension.